### PR TITLE
[WIP] Use `AMPLRepnVisitor` to generate incidence graph in Incidence Analysis

### PR DESCRIPTION
**Pull request by Robbybp**
## Fixes
Potentially unexpected behavior in Incidence Analysis caused (partially) by https://github.com/Pyomo/pyomo/pull/2722. See added tests in test_interface.py.

## Summary/Motivation:
Behavior of Incidence Analysis when terms can cancel out of constraint expressions is not well defined. Currently, `0*x` expressions are removed during expression construction, so they don't appear in the incidence graph. Other "linear cancellations", such as `x - x`, are not removed. https://github.com/Pyomo/pyomo/pull/2722 will remove the `0*x -> 0` optimization, in part because the result is undefined (or our current behavior is incorrect) in odd cases like `0*float('nan')`. This PR uses the `AMPLRepnVisitor` introduced in the new NL writer to more robustly construct the "correct" incidence graph.

## Changes proposed in this PR:
- Use `AMPLRepnVisitor` instead of `identify_variables` to decide which variables to include in an incidence graph. This involves a couple of tricks/hacks to get it to work correctly (according to the current tests). See comments in the code below.
- A small patch to `TemporarySubsystemManager` allowing us to specify variables to temporarily unfix. This is necessary to preserve the current behavior of the `include_fixed` argument to `IncidenceGraphInterface` while updating to the NL visitor.

## Performance
The following is a rough profile of running Dulmage-Mendelsohn on a 35-node dynamic pipeline model I was debugging recently. Here `NFE` is the number of (backward difference) finite elements in both the time domain and the length domain of each pipeline.

Using `main`:
```
NFE = 2
N. var: 5101
N. con: 5101
N. edges: 14530
N. over/under-constrained: 3269
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1     1.215     1.215  100.0
     ------------------------------------------------------
     construct_graph           1     0.135     0.135   11.1
     create_instance           1     0.959     0.959   78.9
     dulmage_mendelsohn        1     0.121     0.121    9.9
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================

NFE = 4
N. var: 12899
N. con: 12899
N. edges: 38134
N. over/under-constrained: 9613
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1     2.468     2.468  100.0
     ------------------------------------------------------
     construct_graph           1     0.241     0.241    9.8
     create_instance           1     1.995     1.995   80.8
     dulmage_mendelsohn        1     0.232     0.232    9.4
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================

NFE = 8
N. var: 39103
N. con: 39103
N. edges: 118798
N. over/under-constrained: 31949
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1     6.841     6.841  100.0
     ------------------------------------------------------
     construct_graph           1     1.015     1.015   14.8
     create_instance           1     5.101     5.101   74.6
     dulmage_mendelsohn        1     0.725     0.725   10.6
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================

NFE = 16
N. var: 133943
N. con: 133943
N. edges: 413950
N. over/under-constrained: 115213
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1    22.484    22.484  100.0
     ------------------------------------------------------
     construct_graph           1     2.608     2.608   11.6
     create_instance           1    16.371    16.371   72.8
     dulmage_mendelsohn        1     3.505     3.505   15.6
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================

NFE = 32
N. var: 493351
N. con: 493351
N. edges: 1539550
N. over/under-constrained: 436109
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1    78.851    78.851  100.0
     ------------------------------------------------------
     construct_graph           1    10.000    10.000   12.7
     create_instance           1    57.667    57.667   73.1
     dulmage_mendelsohn        1    11.184    11.184   14.2
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================
```

Using this PR:
```
NFE = 2
N. var: 5101
N. con: 5101
N. edges: 14506
N. over/under-constrained: 3268
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1     1.362     1.362  100.0
     ------------------------------------------------------
     construct_graph           1     0.317     0.317   23.3
     create_instance           1     0.957     0.957   70.3
     dulmage_mendelsohn        1     0.088     0.088    6.4
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================

NFE = 4
N. var: 12899
N. con: 12899
N. edges: 38094
N. over/under-constrained: 9612
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1     2.956     2.956  100.0
     ------------------------------------------------------
     construct_graph           1     0.705     0.705   23.9
     create_instance           1     1.986     1.986   67.2
     dulmage_mendelsohn        1     0.265     0.265    9.0
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================

NFE = 8
N. var: 39103
N. con: 39103
N. edges: 118726
N. over/under-constrained: 31948
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1     8.450     8.450  100.0
     ------------------------------------------------------
     construct_graph           1     2.479     2.479   29.3
     create_instance           1     5.095     5.095   60.3
     dulmage_mendelsohn        1     0.876     0.876   10.4
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================

NFE = 16
N. var: 133943
N. con: 133943
N. edges: 413814
N. over/under-constrained: 115212
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1    27.960    27.960  100.0
     ------------------------------------------------------
     construct_graph           1     7.511     7.511   26.9
     create_instance           1    16.338    16.338   58.4
     dulmage_mendelsohn        1     4.111     4.111   14.7
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================

NFE = 32
N. var: 493351
N. con: 493351
N. edges: 1539286
N. over/under-constrained: 436108
Identifier                ncalls   cumtime   percall      %
-----------------------------------------------------------
root                           1   101.289   101.289  100.0
     ------------------------------------------------------
     construct_graph           1    29.246    29.246   28.9
     create_instance           1    57.686    57.686   57.0
     dulmage_mendelsohn        1    14.357    14.357   14.2
     other                   n/a     0.000       n/a    0.0
     ======================================================
===========================================================
```

This PR appears to cause a penalty of approximately 3x in constructing the incidence graph for this model. On this particular graph, it also appears to cause a slowdown in Dulmage-Mendelsohn, which I can't explain. In both cases, the dominant computational cost is still model construction (which includes discretization transformations).

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/incidence_analysis/interface.py [hunk 1]
```diff
@@ -101,10 +111,36 @@ def get_bipartite_incidence_graph(variables, constraints, include_fixed=True):
...
+    )
+    to_reset = [
+        var for var in vars_in_cons if var.fixed and var.value is None
+    ]
+    if include_fixed:
+        # AMPLRepnVisitor will not include fixed variables, so if we need to
+        # include these variables, we temporarily unfix them to generate the
+        # AMPLRepn, then re-fix them when we are done.
+        to_unfix = [var for var in variables if var.fixed]
+        context = TemporarySubsystemManager(to_unfix=to_unfix, to_reset=to_reset)
```
Comment (Robbybp) [inline pyomo/contrib/incidence_analysis/interface.py:125]: To preserve the `include_fixed` behavior, we temporarily unfix variables to generate the `AMPLRepn`.

File: pyomo/contrib/incidence_analysis/interface.py [hunk 2]
```diff
@@ -101,10 +111,36 @@ def get_bipartite_incidence_graph(variables, constraints, include_fixed=True):
...
-    for i, con in enumerate(constraints):
-        for var in identify_variables(con.expr, include_fixed=include_fixed):
-            if var in var_node_map:
-                graph.add_edge(i, var_node_map[var])
+    vars_in_cons = list(
+        _generate_variables_in_constraints(constraints, include_fixed=True)
+    )
+    to_reset = [
+        var for var in vars_in_cons if var.fixed and var.value is None
+    ]
```
Comment (Robbybp) [inline pyomo/contrib/incidence_analysis/interface.py:119]: Fixed variables with values of `None` cause problems when we attempt to evaluate them to deterime the constant portion or linear coefficients of the expression. However, the lack of a value for these cofficients/constants should not prevent us from constructing the incidence graph, so we will set them to something other than `None` to generate the `AMPLRepn`, then reset them to `None` afterwards.

File: pyomo/contrib/incidence_analysis/interface.py [hunk 3]
```diff
@@ -101,10 +111,36 @@ def get_bipartite_incidence_graph(variables, constraints, include_fixed=True):
...
+        # is irrelevant for the purpose of an incidence graph, we will
+        # temporarily assign these uninitialized fixed variables a value of
+        # 0.0, then reset them to None after generating the incidence graph.
+        context = TemporarySubsystemManager(to_reset=to_reset)
+
+    with context:
+        for var in to_reset:
+            # We use random.uniform here to defend against arbitrary fixed
+            # variable values causing a numeric cancellation.
+            var.set_value(random.uniform(0.0, 1.0))
```
Comment (Robbybp) [inline pyomo/contrib/incidence_analysis/interface.py:139]: Setting the values of `None`-valued fixed variables can cause us to generate a "wrong" incidence graph if the values we choose happen to cause a numeric cancellation. The obvious values of zero and one happen to be particularly likely to cause a cancellation via terms like `x*y` and `(1-x)*y`. This is a horrible hack but leads to a vanishingly small probability of introducing a numeric cancellation. Alternative suggestions welcome.

**Review by Robbybp — COMMENTED** [review]
See below the tricks/hacks I used to get this to work.

File: pyomo/contrib/incidence_analysis/interface.py (hunk 2 above)
Comment (michaelbynum) [reply pyomo/contrib/incidence_analysis/interface.py:119]: I think `generate_standard_repn` has a `compute_values=False` option which allows for things like this. Does the AMPLRepn have a similar concept?

Comment (Robbybp) [reply pyomo/contrib/incidence_analysis/interface.py:119]: Does `generate_standard_repn` do "linear cancellations", including `0*x -> 0`, with `compute_values=False`?

Comment (michaelbynum) [reply pyomo/contrib/incidence_analysis/interface.py:119]: where `x` is unfixed or fixed to `None`?

Comment (michaelbynum) [reply pyomo/contrib/incidence_analysis/interface.py:119]: Maybe it does not matter.

File: pyomo/contrib/incidence_analysis/interface.py (hunk 3 above)
Comment (Robbybp) [reply pyomo/contrib/incidence_analysis/interface.py:139]: We could get around using `random.uniform` here if `AMPLRepnVisitor` did not replace `NaN*x -> NaN`. In this case, the `AMPLRepn` linear portion would look like `{id(x): float('nan')`. In this case, we could temporarily set fixed variables (and mutable parameters) to NaN to guarantee no cancellation while generating the `AMPLRepn`.

File: pyomo/contrib/incidence_analysis/interface.py (hunk 2 above)
Comment (Robbybp) [reply pyomo/contrib/incidence_analysis/interface.py:119]: I specifically would like it to make this optimization when `x` is unfixed.

Comment (Robbybp) [reply pyomo/contrib/incidence_analysis/interface.py:119]: If `y` is fixed to zero and `compute_values=False`, I would like it to *not* make the optimization `x*y -> 0`. I'll test this behavior and prototype a `generate_standard_repn` backend when I have some time.

Comment (michaelbynum) [reply pyomo/contrib/incidence_analysis/interface.py:119]: I think this is the behavior you are looking for?

```
In [1]: import pyomo.environ as pe

In [2]: from pyomo.repn.standard_repn import generate_standard_repn

In [3]: from pyomo.core.expr.numeric_expr import ProductExpression

In [4]: m = pe.ConcreteModel()

In [5]: m.x = pe.Var()

In [6]: print(generate_standard_repn(ProductExpression((0, m.x)), compute_values=False))

constant:       0
linear vars:    []
linear var ids: []
linear coef:    []
quadratic vars:    []
quadratic var ids: []
quadratic coef:    []
nonlinear expr: None
nonlinear vars: []



In [7]: m.y = pe.Var()

In [8]: m.y.fix(0)

In [9]: print(generate_standard_repn(m.x*m.y, compute_values=False))

constant:       0
linear vars:    ['x']
linear var ids: [4377918416]
linear coef:    [<pyomo.core.base.var.ScalarVar object at 0x1065c59d0>]
quadratic vars:    []
quadratic var ids: []
quadratic coef:    []
nonlinear expr: None
nonlinear vars: []
```

**Comment by mrmundt** [comment]
@Robbybp - FYI, please don't use the GitHub "Draft" feature. It messes with the Jenkins connector.

**Pull request closed without merging**